build-support/scripts/protobuf.sh

```diff
@@ -0,0 +1,244 @@
+readonly FN_DIR="$(dirname "${SCRIPT_DIR}")/functions"
+
+source "${SCRIPT_DIR}/functions.sh"
+
+unset CDPATH
+
+set -euo pipefail
+
+usage() {
+cat <<-EOF
+Usage: ${SCRIPT_NAME} [<options ...>]
+
+Description:
+    Generate all Go files from protobuf definitions. In addition to
+    running the protoc generator it will also fixup build tags in the
+    generated code and regenerate mog outputs and RPC stubs.
+
+Options:
+    --protoc-bin             Path to protoc.
+    --tools-only             Install all required tools but do not generate outputs.
+    -h | --help              Print this help text.
+EOF
+}
+
+err_usage() {
+    err "$1"
+    err ""
+    err "$(usage)"
+}
+
+main() {
+    local protoc_version=
+    local tools_only=
+
+    while test $# -gt 0
+    do
+        case "$1" in
+            -h | --help )
+                usage
+                return 0
+                ;;
+            --protoc-version )
+                protoc_version="$2"
+                shift 2
+                ;;
+            --tools-only )
+                tools_only=1
+                shift
+                ;;
+        esac
+    done
+
+    if test -z "${protoc_version}"
+    then
+        protoc_version="$(make --no-print-directory print-PROTOC_VERSION)"
+        if test -z "${protoc_version}"
+        then
+            err_usage "ERROR: no proto-version specified and version could not be discovered"
+            return 1
+        fi
+    fi
+
+    # ensure the correct protoc compiler is installed
+    protoc_install "${protoc_version}"
+    if test -z "${protoc_bin}" ; then
+        exit 1
+    fi
+
+    # ensure these tools are installed
+    proto_tools_install
+
+    if [[ -n $tools_only ]]; then
+        return 0
+    fi
+
+    declare -a proto_files
+    while IFS= read -r pkg; do
+        pkg="${pkg#"./"}"
+        proto_files+=( "$pkg" )
+    done < <(find . -name '*.proto' | grep -v 'vendor/' | grep -v '.protobuf' | sort )
+
+    for proto_file in "${proto_files[@]}"; do
+        ${SCRIPT_DIR}/proto-gen.sh --grpc --protoc-bin "${protoc_bin}" "$proto_file"
+    done
+
+	echo "Generated all protobuf Go files"
+
+    return 0
+}
+
+# Installs the version of protoc specified by the first argument.
+#
+# Will set 'protoc_bin'
+function protoc_install {
+    local protoc_version="${1:-}"
+    local protoc_os
+
+    if test -z "${protoc_version}"
+    then
+        protoc_version="$(make --no-print-directory print-PROTOC_VERSION)"
+        if test -z "${protoc_version}"
+        then
+            err "ERROR: no protoc-version specified and version could not be discovered"
+            return 1
+        fi
+    fi
+
+    case "$(uname)" in
+        Darwin)
+            protoc_os="osx"
+            ;;
+        Linux)
+            protoc_os="linux"
+            ;;
+        *)
+            err "unexpected OS: $(uname)"
+            return 1
+    esac
+
+    local protoc_zip="protoc-${protoc_version}-${protoc_os}-x86_64.zip"
```

Comment:
This is just lifted up out of the former `make protoc-install` target and decorated with colorized prints courtesy of `functions/10-util.sh` helpers like `status` and `err`